### ISSUE TITLE
Linking error when building libtensorflow_cc.so

### ISSUE THREAD

**Issue by abcdabcd987** [tensorflow/tensorflow#35623]
<em>Please make sure that this is a build/installation issue. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:build_template</em>

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): Source
- TensorFlow version: `v2.1.0-rc2` branch
- Python version: 3.7
- Installed using virtualenv? pip? conda?: No
- Bazel version (if compiling from source): 0.29.1
- GCC/Compiler version (if compiling from source): 7.4.0
- CUDA/cuDNN version: CUDA 10.1 / cuDNN 7.6.5
- GPU model and memory: RTX 2070 8GB

**Describe the problem**

My project uses TensorFlow in C++, so I'm trying to get C++ related `.so` compiled. The problem now is that in the linking stage, there are lots of undefined references, see the following for example and the attached log for more.

```
mlir_passthrough_op.cc:(.text.startup._Z41__static_initialization_and_destruction_0ii.constprop.29+0x2f8): undefined reference to `tensorflow::OpDefBuilder::SetShapeFn(std::function<tensorflow::Status (tensorflow::shape_inference::InferenceContext*)>)'
mlir_passthrough_op.cc:(.text.startup._Z41__static_initialization_and_destruction_0ii.constprop.29+0x324): undefined reference to `tensorflow::register_op::OpDefBuilderReceiver::OpDefBuilderReceiver(tensorflow::register_op::OpDefBuilderWrapper<true> const&)'
bazel-out/k8-opt/bin/tensorflow/stream_executor/cuda/libcuda_stub.pic.a(cuda_stub.pic.o): In function `cudaError_enum (*(anonymous namespace)::LoadSymbol<cudaError_enum (*)(cudaError_enum, char const**)>(char const*))(cudaError_enum, char const**)':
cuda_stub.cc:(.text._ZN12_GLOBAL__N_110LoadSymbolIPF14cudaError_enumS1_PPKcEEET_S3_+0x2a): undefined reference to `tensorflow::Env::Default()'
bazel-out/k8-opt/bin/tensorflow/stream_executor/cuda/libcuda_stub.pic.a(cuda_stub.pic.o): In function `cudaError_enum (*(anonymous namespace)::LoadSymbol<cudaError_enum (*)(int*)>(char const*))(int*)':
cuda_stub.cc:(.text._ZN12_GLOBAL__N_110LoadSymbolIPF14cudaError_enumPiEEET_PKc+0x2a): undefined reference to `tensorflow::Env::Default()'
bazel-out/k8-opt/bin/tensorflow/stream_executor/cuda/libcuda_stub.pic.a(cuda_stub.pic.o): In function `cudaError_enum (*(anonymous namespace)::LoadSymbol<cudaError_enum (*)(char*, int, int)>(char const*))(char*, int, int)':
cuda_stub.cc:(.text._ZN12_GLOBAL__N_110LoadSymbolIPF14cudaError_enumPciiEEET_PKc+0x2a): undefined reference to `tensorflow::Env::Default()'
bazel-out/k8-opt/bin/tensorflow/stream_executor/cuda/libcuda_stub.pic.a(cuda_stub.pic.o): In function `cudaError_enum (*(anonymous namespace)::LoadSymbol<cudaError_enum (*)(int)>(char const*))(int)':
cuda_stub.cc:(.text._ZN12_GLOBAL__N_110LoadSymbolIPF14cudaError_enumiEEET_PKc+0x2a): undefined reference to `tensorflow::Env::Default()'
bazel-out/k8-opt/bin/tensorflow/stream_executor/cuda/libcuda_stub.pic.a(cuda_stub.pic.o): In function `cudaError_enum (*(anonymous namespace)::LoadSymbol<cudaError_enum (*)(CUctx_st*)>(char const*))(CUctx_st*)':
cuda_stub.cc:(.text._ZN12_GLOBAL__N_110LoadSymbolIPF14cudaError_enumP8CUctx_stEEET_PKc+0x2a): undefined reference to `tensorflow::Env::Default()'
bazel-out/k8-opt/bin/tensorflow/stream_executor/cuda/libcuda_stub.pic.a(cuda_stub.pic.o):cuda_stub.cc:(.text._ZN12_GLOBAL__N_110LoadSymbolIPF14cudaError_enumPP8CUctx_stEEET_PKc+0x2a): more undefined references to `tensorflow::Env::Default()' follow
bazel-out/k8-opt/bin/tensorflow/core/distributed_runtime/librequest_id.pic.a(request_id.pic.o): In function `tensorflow::GetUniqueRequestId()':
request_id.cc:(.text._ZN10tensorflow18GetUniqueRequestIdEv+0x9): undefined reference to `tensorflow::random::New64()'
collect2: error: ld returned 1 exit status
INFO: Elapsed time: 33.036s, Critical Path: 32.21s
INFO: 0 processes.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
```

**Provide the exact sequence of commands / steps that you executed before running into the problem**

```
bazel build --config=opt \
                //tensorflow:libtensorflow_cc.so \
                //tensorflow:libtensorflow_framework.so \
                //tensorflow:install_headers
```

.tf_configure.bazelrc:

```
build --action_env PYTHON_BIN_PATH="/home/abcdabcd987/.pyenv/versions/3.7.2/bin/python"
build --action_env PYTHON_LIB_PATH="/home/abcdabcd987/.pyenv/versions/3.7.2/lib/python3.7/site-packages"
build --python_path="/home/abcdabcd987/.pyenv/versions/3.7.2/bin/python"
build:xla --define with_xla_support=true
build --action_env TF_CUDA_VERSION="10.1"
build --action_env TF_CUDNN_VERSION="7"
build --action_env TF_NCCL_VERSION=""
build --action_env TF_CUDA_PATHS="/usr/local/cuda-10.1"
build --action_env CUDA_TOOLKIT_PATH="/usr/local/cuda-10.1"
build --action_env TF_CUDA_COMPUTE_CAPABILITIES="7.5"
build --action_env GCC_HOST_COMPILER_PATH="/usr/bin/x86_64-linux-gnu-gcc-7"
build --config=cuda
build:opt --copt=-march=native
build:opt --copt=-Wno-sign-compare
build:opt --host_copt=-march=native
build:opt --define with_default_optimizations=true
test --flaky_test_attempts=3
test --test_size_filters=small,medium
test --test_tag_filters=-benchmark-test,-no_oss,-oss_serial
test --build_tag_filters=-benchmark-test,-no_oss
test --test_tag_filters=-gpu
test --build_tag_filters=-gpu
build --action_env TF_CONFIGURE_IOS="0"
```

**Any other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

[tensorflow_cc-link.txt.gz](https://github.com/tensorflow/tensorflow/files/4028250/tensorflow_cc-link.txt.gz)

**Comment by Arjuna197** [comment]
Try adding --noincompatible_do_not_split_linking_cmdline to your bazel build command.

**Comment by abcdabcd987** [comment]
Just realized that `bazel` doesn't invoke the correctly versioned one. If I use `/usr/bin/bazel-0.29.1-linux-x86_64`, it works. I ran into this issue because I'm compiling different versions of TensorFlow and have multiple installation of bazel. I also ran into some linking issues when compiling `v1.15.0` and using `/usr/bin/bazel-0.26.1-linux-x86_64` fixed that.

Thanks for the help though @Arjuna197  :)

**Comment by abcdabcd987** [comment]
Actually, I found manually specifying bazel version only fix linking issues for `v1.15.0`, not for 2.x, as in 2.x there is a `.bazelversion` that forces bazel to invoke the specific version. And I found that for 2.x, I can build it on Manjaro (Arch Linux) but not Ubuntu 18.04. `--noincompatible_do_not_split_linking_cmdline` can indeed fix the linking errors on Ubuntu 18.04. Thanks @Arjuna197 !
